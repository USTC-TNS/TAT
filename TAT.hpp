/* TAT/TAT.hpp
 * Copyright (C) 2019  Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef TAT_HPP_
#define TAT_HPP_

#ifndef TAT_VERSION
#define TAT_VERSION "unknown"
#endif // TAT_VERSION

#include <cassert>
#include <cstring>
#include <cmath>
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <vector>
#include <algorithm>
#include <memory>
#include <functional>

#define PASS std::cerr << "calling a passing function at " << __FILE__ << ":" << __LINE__ << " in " << __PRETTY_FUNCTION__ << std::endl, exit(233)
#define ENABLE_IF(...) class = typename std::enable_if<__VA_ARGS__::value>::type

#if (!defined TAT_USE_CPU && !defined TAT_USE_CUDA && !defined TAT_USE_DCU && !defined TAT_USE_SW)
#warning use CPU by default
#define TAT_USE_CPU
#endif

#ifdef TAT_USE_CPU
extern "C"
{
#include <mkl.h>
} // extern "C"
#include <hptt.h>
#include <rang.hpp>

// SVD
#if (defined TAT_USE_GESDD && defined TAT_USE_GESVD) || (defined TAT_USE_GESVD && defined TAT_USE_GESVDX) || (defined TAT_USE_GESVDX && defined TAT_USE_GESDD)
#error only one of GESDD, GESVD and GESVDX could be in use
#endif
#if (!defined TAT_USE_GESDD && !defined TAT_USE_GESVD && !defined TAT_USE_GESVDX)
#warning must use one of GESDD, GESVD and GESVDX, default use GESVD now
#define TAT_USE_GESVD
#endif

// QR
#if (defined TAT_USE_GEQRF && defined TAT_USE_GEQP3)
#error only one of GEQRF and GEQP3 could be in use
#endif
#if (!defined TAT_USE_GEQRF && !defined TAT_USE_GEQP3)
#warning must use one of GEQRF and GEQP3, default use GEQRF now
#define TAT_USE_GEQRF
#endif

#ifdef TAT_USE_GEQP3
#error GEQP3 is current unusable
#endif

#endif // TAT_USE_CPU

namespace TAT {
  enum class Device : unsigned char {CPU, CUDA, DCU, SW};

  namespace legs {
    enum class Legs : unsigned char {
#define CreateLeg(x) \
        Left##x, Right##x, Up##x, Down##x, Phy##x, \
        LeftUp##x, LeftDown##x, RightUp##x, RightDown##x
      CreateLeg(), CreateLeg(1), CreateLeg(2), CreateLeg(3), CreateLeg(4),
      CreateLeg(5), CreateLeg(6), CreateLeg(7), CreateLeg(8), CreateLeg(9)
#undef CreateLeg
    }; // enum class Legs

    inline namespace io {
#define IncEnum(p) {Legs::p, #p}
#define IncGroup(x) \
        IncEnum(Left##x), IncEnum(Right##x), IncEnum(Up##x), IncEnum(Down##x), IncEnum(Phy##x), \
        IncEnum(LeftUp##x), IncEnum(LeftDown##x), IncEnum(RightUp##x), IncEnum(RightDown##x)
      static const std::map<Legs, std::string> legs_str = {
        IncGroup(), IncGroup(1), IncGroup(2), IncGroup(3), IncGroup(4),
        IncGroup(5), IncGroup(6), IncGroup(7), IncGroup(8), IncGroup(9)
      };
#undef IncGroup
#undef IncEnum

      std::ostream& operator<<(std::ostream& out, const Legs& value) {
        return out << legs_str.at(value);
      } // operator<<
    } // namespace io

    inline namespace scalar {
#define IncEnum(p, q) {Legs::p, Legs::q}
#define IncGroup(x) \
        IncEnum(Left##x, Right##x), IncEnum(Right##x, Left##x), IncEnum(Up##x, Down##x), IncEnum(Down##x, Up##x), IncEnum(Phy##x, Phy##x), \
        IncEnum(LeftUp##x, RightDown##x), IncEnum(LeftDown##x, RightUp##x), IncEnum(RightUp##x, LeftDown##x), IncEnum(RightDown##x, LeftUp##x)
      static const std::map<Legs, Legs> minus_legs = {
        IncGroup(), IncGroup(1), IncGroup(2), IncGroup(3), IncGroup(4),
        IncGroup(5), IncGroup(6), IncGroup(7), IncGroup(8), IncGroup(9)
      };
#undef IncGroup
#undef IncEnum

      Legs operator-(const Legs& value) {
        return minus_legs.at(value);
      } // operator-

#define IncEnum(p, q) {Legs::p, Legs::q}
#define IncGroup(x, y) \
        IncEnum(Left##x, Left##y), IncEnum(Right##x, Right##y), IncEnum(Up##x, Up##y), IncEnum(Down##x, Down##y), IncEnum(Phy##x, Phy##y), \
        IncEnum(LeftUp##x, LeftUp##y), IncEnum(LeftDown##x, LeftDown##y), IncEnum(RightUp##x, RightUp##y), IncEnum(RightDown##x, RightDown##y)
      static const std::map<Legs, Legs> plus_legs = {
        IncGroup(, 1), IncGroup(1, 2), IncGroup(2, 3), IncGroup(3, 4), IncGroup(4, 5),
        IncGroup(5, 6), IncGroup(6, 7), IncGroup(7, 8), IncGroup(8, 9), IncGroup(9,)
      };
#undef IncGroup
#undef IncEnum

      Legs operator+(const Legs& value) {
        return plus_legs.at(value);
      } // operator+
    } // namespace scalar;
  } // namespace legs
  using legs::Legs;

  namespace legs_name {
#define TAT_DefineLeg(x) static const TAT::Legs x = TAT::Legs::x
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Left##n); TAT_DefineLeg(Right##n); TAT_DefineLeg(Up##n); TAT_DefineLeg(Down##n); TAT_DefineLeg(Phy##n); \
      TAT_DefineLeg(LeftUp##n); TAT_DefineLeg(LeftDown##n); TAT_DefineLeg(RightUp##n); TAT_DefineLeg(RightDown##n)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)

    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#undef TAT_DefineLeg
  } // namespace legs_name

  using Size = std::size_t;
  using Rank = unsigned int;

  namespace data {
    template<Device device, class Base>
    class Magic;
#define DefineData(x) \
      namespace x { \
        template<class Base> \
        class Data; \
      } \
      template<class Base> \
      class Magic<Device::x, Base>{ \
       public: \
        using type=x::Data<Base>; \
      }

    DefineData(CPU);
    DefineData(CUDA);
    DefineData(DCU);
    DefineData(SW);
#undef DefineData

    template<Device device, class Base, ENABLE_IF(std::is_scalar<Base>)>
    using Data = typename Magic<device, Base>::type;
  } // namespace data
  using data::Data;

  namespace node {
    template<Device device, class Base>
    class Node;
  } // namespace node
  using node::Node;

  namespace tensor {
    template<Device device=Device::CPU, class Base=double>
    class Tensor;
  } // namespace tensor
  using tensor::Tensor;

  namespace site {
    template<Device device=Device::CPU, class Base=double>
    class Site;
  } // namespace site
  using site::Site;

  namespace lattice {
    template<int dimension>
    class Dimension;

    template<class Tags=Dimension<2>, Device device=Device::CPU, class Base=double>
    class Lattice;
  } // namespace lattice
  using site::Site;
} // namespace TAT

#include "Data.hpp"
#include "Node.hpp"
#include "Tensor.hpp"
#include "Site.hpp"
#include "Lattice.hpp"

#endif // TAT_HPP_
