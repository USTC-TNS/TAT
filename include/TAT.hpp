/* TAT.hpp
 * Copyright (C) 2019  Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef TAT_HPP_
#define TAT_HPP_

#ifndef TAT_VERSION
#define TAT_VERSION "unknown"
#endif // TAT_VERSION

#include <cassert>
#include <cstring>
#include <cmath>
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <vector>
#include <algorithm>
#include <memory>
#include <functional>
#include <complex>

#define PASS std::cerr << "calling a passing function at " << __FILE__ << ":" << __LINE__ << " in " << __PRETTY_FUNCTION__ << std::endl, exit(233)
#define ENABLE_IF(...) class = typename std::enable_if<__VA_ARGS__::value>::type

#if (!defined TAT_USE_CPU && !defined TAT_USE_CUDA && !defined TAT_USE_DCU && !defined TAT_USE_SW)
#if !defined TAT_DEFAULT
#warning use CPU by default
#endif
#define TAT_USE_CPU
#endif

#ifdef TAT_EXTREME
#warning EXTREME compile may cost much of compile time
#endif // TAT_EXTREME

#ifdef TAT_USE_CPU
extern "C"
{
#define MKL_Complex8 std::complex<float>
#define MKL_Complex16 std::complex<double>
#include <mkl.h>
} // extern "C"
#include <hptt.h>
#ifdef TAT_EXTREME
#include <../src/hptt.cpp>
#include <../src/plan.cpp>
#include <../src/transpose.cpp>
#include <../src/utils.cpp>
#endif // TAT_EXTREME
#include <rang.hpp>

// SVD
#if (defined TAT_USE_GESDD && defined TAT_USE_GESVD) || (defined TAT_USE_GESVD && defined TAT_USE_GESVDX) || (defined TAT_USE_GESVDX && defined TAT_USE_GESDD)
#error only one of GESDD, GESVD and GESVDX could be in use
#endif
#if (!defined TAT_USE_GESDD && !defined TAT_USE_GESVD && !defined TAT_USE_GESVDX)
#if !defined TAT_DEFAULT
#warning must use one of GESDD, GESVD and GESVDX, default use GESVD now
#endif
#define TAT_USE_GESVD
#endif

// QR
#if (defined TAT_USE_GEQRF && defined TAT_USE_GEQP3)
#error only one of GEQRF and GEQP3 could be in use
#endif
#if (!defined TAT_USE_GEQRF && !defined TAT_USE_GEQP3)
#if !defined TAT_DEFAULT
#warning must use one of GEQRF and GEQP3, default use GEQRF now
#endif
#define TAT_USE_GEQRF
#endif

#ifdef TAT_USE_GEQP3
#error GEQP3 is current unusable
#endif

#endif // TAT_USE_CPU

namespace TAT {
  template<class T>
  class is_scalar {
   public:
    static constexpr bool value = std::is_scalar<T>::value || std::is_same<T, std::complex<float>>::value || std::is_same<T, std::complex<double>>::value || std::is_same<T, std::complex<long double>>::value;
  };

  template<class Base>
  class RealBaseClass {
   public:
    using type=Base;
  };
  template<>
  class RealBaseClass<std::complex<float>> {
   public:
    using type=float;
  };
  template<>
  class RealBaseClass<std::complex<double>> {
   public:
    using type=double;
  };
  template<>
  class RealBaseClass<std::complex<long double>> {
   public:
    using type=long double;
  };
  template<class T>
  using RealBase = typename RealBaseClass<T>::type;

  enum class Device : unsigned char {CPU, CUDA, DCU, SW};

  namespace legs {
    class Legs {
     public:
      unsigned char id;
      bool operator==(const Legs& other) const {
        return id==other.id;
      }
      bool operator!=(const Legs& other) const {
        return id!=other.id;
      }
      bool operator<(const Legs& other) const {
        return id<other.id;
      }
      Legs() = default;
      Legs(const std::string& name) {
        try {
          id = name2id.at(name);
        } catch (const std::out_of_range& e) {
          id = total++;
          name2id[name]=id;
          id2name[id]=name;
        } // exsit name
      }
      static unsigned char total;
      static std::map<std::string, unsigned char> name2id;
      static std::map<unsigned char, std::string> id2name;

      friend std::ostream& operator<<(std::ostream& out, const Legs& value) {
        try {
          return out << id2name.at(value.id);
        } catch (const std::out_of_range& e) {
          return out << "UserDefinedLeg" << value.id;
        } // get name
      } // operator<<
    }; // class Legs

    unsigned char Legs::total = 0;
    std::map<std::string, unsigned char> Legs::name2id = {};
    std::map<unsigned char, std::string> Legs::id2name = {};
  } // namespace legs
  using legs::Legs;

  namespace legs_name {
#define TAT_DefineLeg(x) static const TAT::Legs x(#x)
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Phy##n); TAT_DefineLeg(Left##n); TAT_DefineLeg(Right##n); TAT_DefineLeg(Up##n); TAT_DefineLeg(Down##n); \
      TAT_DefineLeg(LeftUp##n); TAT_DefineLeg(LeftDown##n); TAT_DefineLeg(RightUp##n); TAT_DefineLeg(RightDown##n)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Tmp##n##0); TAT_DefineLeg(Tmp##n##1); TAT_DefineLeg(Tmp##n##2); TAT_DefineLeg(Tmp##n##3); TAT_DefineLeg(Tmp##n##4); \
      TAT_DefineLeg(Tmp##n##5); TAT_DefineLeg(Tmp##n##6); TAT_DefineLeg(Tmp##n##7); TAT_DefineLeg(Tmp##n##8); TAT_DefineLeg(Tmp##n##9)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#undef TAT_DefineLeg
    // total = 190 currently
  } // namespace legs_name

  using Size = std::size_t;
  using Rank = unsigned int;

  namespace data {
    template<Device device, class Base>
    class Magic;
#define DefineData(x) \
      namespace x { \
        template<class Base> \
        class Data; \
      } \
      template<class Base> \
      class Magic<Device::x, Base>{ \
       public: \
        using type=x::Data<Base>; \
      }

    DefineData(CPU);
    DefineData(CUDA);
    DefineData(DCU);
    DefineData(SW);
#undef DefineData

    template<Device device=Device::CPU, class Base=double, ENABLE_IF(is_scalar<Base>)>
    using Data = typename Magic<device, Base>::type;
  } // namespace data
  using data::Data;

  namespace block {
    template<Device device=Device::CPU, class Base=double>
    class Block;
  } // namespace block
  using block::Block;

  namespace tensor {
    template<Device device=Device::CPU, class Base=double>
    class Tensor;
  } // namespace tensor
  //using tensor::Tensor;
  template<Device device=Device::CPU, class Base=double>
  using Tensor=Block<device, Base>;

  namespace node {
    template<Device device=Device::CPU, class Base=double>
    class Node;
  } // namespace node
  using node::Node;

  namespace lazy {
    template<Device device=Device::CPU, class Base=double>
    class Lazy;
  } // namespace lazy
  using lazy::Lazy;

  namespace site {
    template<Device device=Device::CPU, class Base=double>
    class Site;
  } // namespace site
  using site::Site;

  namespace lattice {
    template<int dimension>
    class Dimension;

    template<class Tags=Dimension<2>, Device device=Device::CPU, class Base=double>
    class Lattice;
  } // namespace lattice
  using site::Site;
} // namespace TAT

#include "TAT/Data.hpp"
#include "TAT/Block.hpp"
// Tensor
#include "TAT/Node.hpp"
#include "TAT/Lazy.hpp"

#include "TAT/Site.hpp"
#include "TAT/Lattice.hpp"

#endif // TAT_HPP_
