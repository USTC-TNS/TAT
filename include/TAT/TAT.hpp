/**
 * \file TAT.hpp
 *
 * Copyright (C) 2019-2020 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once
#ifndef TAT_HPP
#define TAT_HPP

#ifndef __cplusplus
#error only work for c++
#endif

#ifdef _MSVC_LANG
#if _MSVC_LANG < 201703L
#error require c++17 or later
#endif
#else
#if __cplusplus < 201703L
#error require c++17 or later
#endif
#endif

#ifdef _WIN32
#include <windows.h>
#endif

// 开关说明
// TAT_USE_MPI 定义以开启MPI支持, cmake可对此进行定义
// TAT_USE_SIMPLE_NAME 定义以使用原始字符串作为name

namespace TAT {
   /**
    * \brief TAT的版本号
    */
   inline const char* version = "0.0.5";

   /**
    * \brief Debug模式中, 将在程序末尾打印一行友情提示, 过早的优化是万恶之源, 同时控制windows下终端的色彩模式
    */
   struct evil_t {
      evil_t() {
#ifdef _WIN32
         HANDLE output_handle = GetStdHandle(STD_OUTPUT_HANDLE);
         DWORD output_mode = 0;
         GetConsoleMode(output_handle, &output_mode);
         SetConsoleMode(output_handle, output_mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
         HANDLE error_handle = GetStdHandle(STD_ERROR_HANDLE);
         DWORD error_mode = 0;
         GetConsoleMode(error_handle, &error_mode);
         SetConsoleMode(error_handle, error_mode | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
#endif
      }
      ~evil_t();
   };
   inline const evil_t evil;
} // namespace TAT

// clang-format off
#include "tensor.hpp"
#include "implement.hpp"
#include "tools.hpp"
// clang-format on

#endif
