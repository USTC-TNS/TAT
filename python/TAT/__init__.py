import cppyy
cppyy.load_library("blas")
cppyy.load_library("lapack")
cppyy.include("../include/TAT/TAT.hpp")
std = cppyy.gbl.std
TAT = cppyy.gbl.TAT
