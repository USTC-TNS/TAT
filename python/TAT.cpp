#include <sstream>

#include <pybind11/functional.h>
#include <pybind11/operators.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#define TAT_ALWAYS_COLOR
#include "TAT/TAT.hpp"

namespace TAT {
   namespace py = pybind11;

   template<class Type, class Args>
   auto implicit_init() {
      py::implicitly_convertible<Args, Type>();
      return py::init<Args>();
   }

   template<class Type, class Args, class Func>
   auto implicit_init(Func&& func) {
      py::implicitly_convertible<Args, Type>();
      return py::init(func);
   }

   template<class ScalarType, class Symmetry>
   auto declare_tensor(py::module& m, const char* name) {
      using T = Tensor<ScalarType, Symmetry>;
      using E = Edge<Symmetry>;
      return py::class_<T>(m, name)
            .def(py::self + py::self)
            .def(ScalarType() + py::self)
            .def(py::self + ScalarType())
            .def(py::self += py::self)
            .def(py::self += ScalarType())
            .def(py::self - py::self)
            .def(ScalarType() - py::self)
            .def(py::self - ScalarType())
            .def(py::self -= py::self)
            .def(py::self -= ScalarType())
            .def(py::self * py::self)
            .def(ScalarType() * py::self)
            .def(py::self * ScalarType())
            .def(py::self *= py::self)
            .def(py::self *= ScalarType())
            .def(py::self / py::self)
            .def(ScalarType() / py::self)
            .def(py::self / ScalarType())
            .def(py::self /= py::self)
            .def(py::self /= ScalarType())
            .def("__str__",
                 [](const T& tensor) {
                    auto out = std::stringstream();
                    out << tensor;
                    return out.str();
                 })
            .def("__repr__",
                 [](const T& tensor) {
                    auto out = std::stringstream();
                    out << "Tensor";
                    out << tensor;
                    return out.str();
                 })
            .def("shape",
                 [](const T& tensor) {
                    auto out = std::stringstream();
                    out << '{' << console_green << "names" << console_origin << ':';
                    out << tensor.names;
                    out << ',' << console_green << "edges" << console_origin << ':';
                    out << tensor.core->edges;
                    out << '}';
                    return out.str();
                 })
            .def(py::init<const std::vector<Name>&, const std::vector<E>&>())
            .def(implicit_init<T, std::pair<std::vector<Name>, std::vector<E>>>([](const std::pair<std::vector<Name>, std::vector<E>>& p) {
               return std::make_unique<T>(p.first, p.second);
            }))
            // TODO problem: `TAT.TensorDoubleU1Symmetry(list("AB"),[{0:2},{0:2}])` not work
            // https://github.com/pybind/pybind11/issues/2182
            .def(implicit_init<T, ScalarType>())
            .def("value", [](const T& tensor) -> ScalarType { return tensor; })
            .def("copy", &T::copy)
            .def("same_shape", &T::same_shape)
            .def("map", [](const T& tensor, std::function<ScalarType(ScalarType)>& function) { return tensor.map(function); })
            .def("transform", [](T& tensor, std::function<ScalarType(ScalarType)>& function) -> T& { return tensor.transform(function); })
            .def("set", [](T& tensor, std::function<ScalarType()>& function) -> T& { return tensor.set(function); })
            .def("zero", [](T& tensor) -> T& { return tensor.zero(); })
            .def(
                  "test",
                  [](T& tensor, ScalarType first, ScalarType step) -> T& { return tensor.test(first, step); },
                  py::arg("first") = 0,
                  py::arg("step") = 1)
            // TODO block and at
            // TODO to
            .def("norm_max", &T::template norm<-1>)
            .def("norm_num", &T::template norm<0>)
            .def("norm_1", &T::template norm<1>)
            .def("norm_2", &T::template norm<2>);
      // TODO edge contract svd
   }

   template<class Symmetry>
   auto declare_edge(py::module& m, const char* name) {
      auto result = py::class_<Edge<Symmetry>>(m, name)
                          .def_readwrite("map", &Edge<Symmetry>::map)
                          .def(implicit_init<Edge<Symmetry>, Size>())
                          .def(implicit_init<Edge<Symmetry>, std::map<Symmetry, Size>>())
                          .def(implicit_init<Edge<Symmetry>, const std::set<Symmetry>&>())
                          .def("__str__",
                               [](const Edge<Symmetry>& edge) {
                                  auto out = std::stringstream();
                                  out << edge;
                                  return out.str();
                               })
                          .def("__repr__", [](const Edge<Symmetry>& edge) {
                             auto out = std::stringstream();
                             out << "Edge";
                             out << edge;
                             return out.str();
                          });
      if constexpr (is_fermi_symmetry_v<Symmetry>) {
         result = result.def_readwrite("arrow", &Edge<Symmetry>::arrow).def(py::init<Arrow, std::map<Symmetry, Size>>());
      }
      return result;
   }

   PYBIND11_MODULE(TAT, m) {
      m.doc() = "TAT is A Tensor library!";
      // name
      py::class_<Name>(m, "Name")
            .def(implicit_init<Name, const std::string&>())
            .def("__repr__", [](const Name& name) { return "Name[" + id_to_name.at(name.id) + "]"; })
            .def("__str__", [](const Name& name) { return id_to_name.at(name.id); })
            .def_readonly("id", &Name::id)
            .def_property_readonly("name", [](const Name& name) { return id_to_name.at(name.id); });

      // symmetry
      m.attr("Symmetry") = py::dict(
            py::arg("NoSymmetry") = py::class_<NoSymmetry>(m, "NoSymmetry")
                                          .def(py::init<>())
                                          .def("__repr__", [=](const NoSymmetry& symmetry) { return std::string("NoSymmetry") + "[" + +"]"; })
                                          .def("__str__",
                                               [=](const NoSymmetry& symmetry) {
                                                  auto out = std::stringstream();
                                                  out << symmetry;
                                                  return out.str();
                                               }),
            py::arg("Z2Symmetry") =
                  py::class_<Z2Symmetry>(m, "Z2Symmetry")
                        .def(implicit_init<Z2Symmetry, Z2>())
                        .def("__repr__",
                             [=](const Z2Symmetry& symmetry) { return std::string("Z2Symmetry") + "[" + std::to_string(symmetry.z2) + "]"; })
                        .def("__str__",
                             [=](const Z2Symmetry& symmetry) {
                                auto out = std::stringstream();
                                out << symmetry;
                                return out.str();
                             })
                        .def_readwrite("z2", &Z2Symmetry::z2),
            py::arg("U1Symmetry") =
                  py::class_<U1Symmetry>(m, "U1Symmetry")
                        .def(implicit_init<U1Symmetry, U1>())
                        .def("__repr__",
                             [=](const U1Symmetry& symmetry) { return std::string("U1Symmetry") + "[" + std::to_string(symmetry.u1) + "]"; })
                        .def("__str__",
                             [=](const U1Symmetry& symmetry) {
                                auto out = std::stringstream();
                                out << symmetry;
                                return out.str();
                             })
                        .def_readwrite("u1", &U1Symmetry::u1),
            py::arg("FermiSymmmetry") =
                  py::class_<FermiSymmetry>(m, "FermiSymmetry")
                        .def(implicit_init<FermiSymmetry, Fermi>())
                        .def("__repr__",
                             [=](const FermiSymmetry& symmetry) { return std::string("FermiSymmetry") + "[" + std::to_string(symmetry.fermi) + "]"; })
                        .def("__str__",
                             [=](const FermiSymmetry& symmetry) {
                                auto out = std::stringstream();
                                out << symmetry;
                                return out.str();
                             })
                        .def_readwrite("fermi", &FermiSymmetry::fermi),
            py::arg("FermiZ2Symmmetry") = py::class_<FermiZ2Symmetry>(m, "FermiZ2Symmetry")
                                                .def(py::init<Fermi, Z2>())
                                                .def(implicit_init<FermiZ2Symmetry, const std::pair<Fermi, Z2>&>([](const std::pair<Fermi, Z2>& p) {
                                                   return std::make_unique<FermiZ2Symmetry>(p.first, p.second);
                                                }))
                                                .def("__repr__",
                                                     [=](const FermiZ2Symmetry& symmetry) {
                                                        return std::string("FermiZ2Symmetry") + "[" + std::to_string(symmetry.fermi) + "," +
                                                               std::to_string(symmetry.z2) + "]";
                                                     })
                                                .def("__str__",
                                                     [=](const FermiZ2Symmetry& symmetry) {
                                                        auto out = std::stringstream();
                                                        out << symmetry;
                                                        return out.str();
                                                     })
                                                .def_readwrite("fermi", &FermiZ2Symmetry::fermi)
                                                .def_readwrite("z2", &FermiZ2Symmetry::z2),
            py::arg("FermiU1Symmmetry") = py::class_<FermiU1Symmetry>(m, "FermiU1Symmetry")
                                                .def(py::init<Fermi, U1>())
                                                .def(implicit_init<FermiU1Symmetry, const std::pair<Fermi, U1>&>([](const std::pair<Fermi, U1>& p) {
                                                   return std::make_unique<FermiU1Symmetry>(p.first, p.second);
                                                }))
                                                .def("__repr__",
                                                     [=](const FermiU1Symmetry& symmetry) {
                                                        return std::string("FermiU1Symmetry") + "[" + std::to_string(symmetry.fermi) + "," +
                                                               std::to_string(symmetry.u1) + "]";
                                                     })
                                                .def("__str__",
                                                     [=](const FermiU1Symmetry& symmetry) {
                                                        auto out = std::stringstream();
                                                        out << symmetry;
                                                        return out.str();
                                                     })
                                                .def_readwrite("fermi", &FermiU1Symmetry::fermi)
                                                .def_readwrite("u1", &FermiU1Symmetry::u1));
      // edge
      m.attr("Edge") = py::dict(
#define DECLARE_EDGE(S) py::arg(#S) = declare_edge<S>(m, "Edge" #S)
            DECLARE_EDGE(NoSymmetry),
            DECLARE_EDGE(Z2Symmetry),
            DECLARE_EDGE(U1Symmetry),
            DECLARE_EDGE(FermiSymmetry),
            DECLARE_EDGE(FermiZ2Symmetry),
            DECLARE_EDGE(FermiU1Symmetry)
#undef DECLARE_EDGE
      );
      // tensor
      declare_tensor<double, NoSymmetry>(m, "TensorDoubleNoSymmetry");
      declare_tensor<double, U1Symmetry>(m, "TensorDoubleU1Symmetry");
      // TODO: python support
   }
} // namespace TAT
