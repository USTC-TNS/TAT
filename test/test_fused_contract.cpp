/**
 * Copyright (C) 2019-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <TAT/TAT.hpp>

#include "run_test.hpp"

using namespace TAT;

void run_test() {
   auto a = Tensor<>({"A", "B", "C"}, {2, 3, 5}).range();
   auto b = Tensor<>({"A", "B", "D"}, {2, 3, 7}).range();
   auto c = Tensor<>::contract(a, b, {{"B", "B"}});
   auto a0 = a.shrink({{"A", 0}});
   auto a1 = a.shrink({{"A", 1}});
   auto b0 = b.shrink({{"A", 0}});
   auto b1 = b.shrink({{"A", 1}});
   auto c0 = c.shrink({{"A", 0}});
   auto c1 = c.shrink({{"A", 1}});
   std::cout << a0.contract(b0, {{"B", "B"}}) - c0 << "\n";
   std::cout << a1.contract(b1, {{"B", "B"}}) - c1 << "\n";
}
