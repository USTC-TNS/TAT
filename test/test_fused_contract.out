{[32mnames[0m:[C,D],[32medges[0m:[5,7],[32mblocks[0m:[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}
{[32mnames[0m:[C,D],[32medges[0m:[5,7],[32mblocks[0m:[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}
